File: scripts/fetch-rhizome.js
#!/usr/bin/env node

/**
 * Fetch Rhizome Community Listings
 *
 * This script uses Playwright to bypass Cloudflare protection and scrape
 * community listings from rhizome.org/community/
 *
 * Requirements: npm install playwright && npx playwright install chromium
 *
 * Usage: node scripts/fetch-rhizome.js
 */

const fs = require('fs');
const path = require('path');

async function fetchRhizomeListings() {
  let browser;

  try {
    // Check if playwright is installed
    let playwright;
    try {
      playwright = require('playwright');
    } catch (err) {
      console.error('Playwright not installed. Run: npm install playwright && npx playwright install chromium');
      process.exit(1);
    }

    // Launch browser
    browser = await playwright.chromium.launch({
      headless: true,
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    });

    const context = await browser.newContext({
      userAgent: 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    });

    const page = await context.newPage();

    console.log('Fetching https://rhizome.org/community/ ...');
    await page.goto('https://rhizome.org/community/', {
      waitUntil: 'networkidle',
      timeout: 30000
    });

    // Wait for Cloudflare challenge to complete
    await page.waitForTimeout(5000);

    // Extract listings
    const listings = await page.evaluate(() => {
      const items = [];

      // Adjust selectors based on actual Rhizome page structure
      const listingSelectors = [
        '.listing-item',
        '.community-listing',
        '.opportunity-item',
        'article.post',
        '.event-item'
      ];

      let elements = [];
      for (const selector of listingSelectors) {
        const found = document.querySelectorAll(selector);
        if (found.length > 0) {
          elements = Array.from(found);
          break;
        }
      }

      // Fallback: look for any article or list items in main content
      if (elements.length === 0) {
        elements = Array.from(document.querySelectorAll('main article, main li, .content article'));
      }

      elements.slice(0, 15).forEach(element => {
        const titleEl = element.querySelector('h1, h2, h3, h4, .title, .listing-title');
        const descEl = element.querySelector('p, .description, .excerpt, .summary');
        const linkEl = element.querySelector('a');
        const dateEl = element.querySelector('time, .date, .published');

        if (titleEl && titleEl.textContent.trim()) {
          const listing = {
            title: titleEl.textContent.trim(),
            description: descEl ? descEl.textContent.trim().substring(0, 200) : '',
            url: linkEl ? new URL(linkEl.href, 'https://rhizome.org').href : 'https://rhizome.org/community/',
            type: element.className || 'listing',
            date: dateEl ? dateEl.getAttribute('datetime') || dateEl.textContent : new Date().toISOString(),
            source: 'Rhizome Community'
          };
          items.push(listing);
        }
      });

      return items;
    });

    // Create data structure
    const data = {
      community_listings: listings,
      last_updated: new Date().toISOString(),
      status: listings.length > 0 ? 'success' : 'no_listings_found',
      url: 'https://rhizome.org/community/',
      note: `Fetched ${listings.length} listings using browser automation`
    };

    // Write to file
    const dataDir = path.join(__dirname, '..', 'data');
    if (!fs.existsSync(dataDir)) {
      fs.mkdirSync(dataDir, { recursive: true });
    }

    const outputPath = path.join(dataDir, 'rhizome.json');
    fs.writeFileSync(outputPath, JSON.stringify(data, null, 2));

    console.log(`✓ Successfully fetched ${listings.length} listings`);
    console.log(`✓ Saved to ${outputPath}`);

    await browser.close();
    process.exit(0);

  } catch (error) {
    console.error('Error fetching Rhizome listings:', error.message);

    // Write error state
    const errorData = {
      community_listings: [],
      last_updated: new Date().toISOString(),
      status: 'error',
      error: error.message,
      url: 'https://rhizome.org/community/',
      note: 'Failed to fetch listings - see error field'
    };

    const outputPath = path.join(__dirname, '..', 'data', 'rhizome.json');
    fs.writeFileSync(outputPath, JSON.stringify(errorData, null, 2));

    if (browser) await browser.close();
    process.exit(1);
  }
}

// Run if called directly
if (require.main === module) {
  fetchRhizomeListings();
}

module.exports = { fetchRhizomeListings };

File: fetch_rhizome.py
#!/usr/bin/env python3
"""
Fetch Rhizome community listings.
Supports both Playwright (for Cloudflare bypass) and fallback requests.
"""

import json
import os
import sys
import re
from datetime import datetime, timezone

def create_fallback_data():
    """Create fallback data structure when scraping is blocked."""
    return {
        "community_listings": [
            {
                "title": "Rhizome Community Access Limited",
                "description": "Cloudflare protection prevents automated scraping. Requires Playwright with browser installation.",
                "url": "https://rhizome.org/community/",
                "type": "system-message",
                "date": datetime.now(timezone.utc).isoformat()
            }
        ],
        "last_updated": datetime.now(timezone.utc).isoformat(),
        "status": "cloudflare-protected",
        "note": "Install: pip install playwright && playwright install chromium"
    }

def parse_community_listings(html_content):
    """Parse HTML content to extract community listings."""
    listings = []

    # Try to extract listing items from common patterns
    # Rhizome uses various structures, so we'll try multiple patterns

    # Pattern 1: Look for article or list items with links
    article_pattern = r'<article[^>]*>(.*?)</article>'
    item_pattern = r'<li[^>]*class="[^"]*community[^"]*"[^>]*>(.*?)</li>'

    # Pattern 2: Look for headings with links
    heading_link_pattern = r'<h[2-4][^>]*>.*?<a[^>]*href="([^"]*)"[^>]*>(.*?)</a>.*?</h[2-4]>'

    # Pattern 3: Look for description paragraphs
    desc_pattern = r'<p[^>]*class="[^"]*description[^"]*"[^>]*>(.*?)</p>'

    # Simple extraction - find all links in the community section
    links = re.findall(r'<a[^>]*href="(/community/[^"]*)"[^>]*>(.*?)</a>', html_content, re.DOTALL)

    for url, title in links[:15]:  # Limit to 15 items
        # Clean up title
        title_clean = re.sub(r'<[^>]+>', '', title).strip()
        if title_clean and len(title_clean) > 3:
            listing = {
                "title": title_clean,
                "description": "",
                "url": f"https://rhizome.org{url}" if url.startswith('/') else url,
                "type": "community-listing",
                "date": datetime.now(timezone.utc).isoformat()
            }
            listings.append(listing)

    return listings

def fetch_with_playwright():
    """Fetch using Playwright to bypass Cloudflare."""
    try:
        from playwright.sync_api import sync_playwright

        print("Using Playwright to bypass Cloudflare...", file=sys.stderr)

        with sync_playwright() as p:
            browser = p.chromium.launch(headless=True)
            page = browser.new_page(
                user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
            )

            # Navigate and wait for page to load
            page.goto('https://rhizome.org/community/', wait_until='networkidle', timeout=60000)

            # Wait a bit for any dynamic content
            page.wait_for_timeout(3000)

            html_content = page.content()
            browser.close()

            # Parse the content
            listings = parse_community_listings(html_content)

            if not listings:
                # If parsing failed, at least we got the page
                return {
                    "community_listings": [{
                        "title": "Rhizome Community",
                        "description": "Successfully accessed page but parsing needs refinement",
                        "url": "https://rhizome.org/community/",
                        "type": "placeholder",
                        "date": datetime.now(timezone.utc).isoformat()
                    }],
                    "last_updated": datetime.now(timezone.utc).isoformat(),
                    "status": "partial-success",
                    "note": "Page accessed via Playwright, HTML parsing may need updates"
                }

            return {
                "community_listings": listings,
                "last_updated": datetime.now(timezone.utc).isoformat(),
                "status": "success"
            }

    except ImportError:
        print("Playwright not installed. Install with: pip install playwright && playwright install chromium", file=sys.stderr)
        return None
    except Exception as e:
        print(f"Playwright error: {e}", file=sys.stderr)
        return None

def fetch_with_requests():
    """Fallback: try with requests (likely to be blocked by Cloudflare)."""
    try:
        import requests

        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        }

        response = requests.get('https://rhizome.org/community/', headers=headers, timeout=30)

        if 'challenge-platform' in response.text or response.status_code == 403:
            print("Cloudflare protection detected with requests.", file=sys.stderr)
            return None

        listings = parse_community_listings(response.text)

        if listings:
            return {
                "community_listings": listings,
                "last_updated": datetime.now(timezone.utc).isoformat(),
                "status": "success"
            }

        return None

    except Exception as e:
        print(f"Requests error: {e}", file=sys.stderr)
        return None

def main():
    # Try Playwright first, then requests, then fallback
    data = fetch_with_playwright()

    if not data:
        print("Trying with requests library...", file=sys.stderr)
        data = fetch_with_requests()

    if not data:
        print("All methods failed, using fallback data.", file=sys.stderr)
        data = create_fallback_data()

    # Ensure data directory exists
    os.makedirs('data', exist_ok=True)

    # Write JSON file
    with open('data/rhizome.json', 'w') as f:
        json.dump(data, f, indent=2)

    print(f"✓ Created data/rhizome.json - Status: {data.get('status', 'unknown')}", file=sys.stderr)
    print(f"  Found {len(data.get('community_listings', []))} listing(s)", file=sys.stderr)

    return 0

if __name__ == '__main__':
    sys.exit(main())
